core/lib/server/pub.ts

```diff
@@ -601,6 +606,383 @@ export const deletePub = async (pubId: PubsId) =>
+						.where("relatedPubId", "is not", null)
+						// upsert
+						.doUpdateSet((eb) => ({
+							value: eb.ref("excluded.value"),
+						}))
+				)
+		).execute();
+
+		return createdRelations;
+	});
+};
+
+/**
+ * Removes specific pub relations by deleting pub_values entries that match the provided relations.
+ * Each relation must specify a field slug and relatedPubId to identify which relation to remove.
+ */
+export const removePubRelations = async ({
+	pubId,
+	relations,
+	communityId,
+	trx = db,
+}: {
+	pubId: PubsId;
+	relations: RemovePubRelationsInput[];
+	communityId: CommunitiesId;
+	trx?: typeof db;
+}) => {
+	const consolidatedRelations = await consolidateFieldSlugs(relations, communityId);
+
+	const removed = await autoRevalidate(
+		trx
+			.deleteFrom("pub_values")
+			.where("pubId", "=", pubId)
+			.where((eb) =>
+				eb.or(
+					consolidatedRelations.map(({ relatedPubId, fieldId }) =>
+						eb.and([eb("fieldId", "=", fieldId), eb("relatedPubId", "=", relatedPubId)])
+					)
+				)
+			)
+			.returning(["pub_values.relatedPubId"])
+	).execute();
+
+	return removed.map(({ relatedPubId }) => relatedPubId);
+};
+
+/**
+ * Removes all relations for a given field slug and pubId
+ */
+export const removeAllPubRelationsBySlugs = async ({
+	pubId,
+	slugs,
+	communityId,
+	trx = db,
+}: {
+	pubId: PubsId;
+	slugs: string[];
+	communityId: CommunitiesId;
+	trx?: typeof db;
+}) => {
+	const fields = await consolidateFieldSlugs(
+		slugs.map((slug) => ({ slug })),
+		communityId
+	);
+	const fieldIds = fields.map(({ fieldId }) => fieldId);
+	if (!fieldIds.length) {
+		throw new Error(`No fields found for slugs: ${slugs.join(", ")}`);
+	}
+
+	const removed = await autoRevalidate(
+		trx
+			.deleteFrom("pub_values")
+			.where("pubId", "=", pubId)
+			.where("fieldId", "in", fieldIds)
+			.where("relatedPubId", "is not", null)
+			.returning("relatedPubId")
+	).execute();
+
+	return removed.map(({ relatedPubId }) => relatedPubId);
+};
+
+/**
+ * Replaces all relations for given field slugs with new relations.
+ * First removes all existing relations for the provided slugs, then adds the new relations.
+ *
+ * If the `relations` object is empty, this function does nothing.
+ */
+export const replacePubRelationsBySlug = async ({
+	pubId,
+	relations,
+	communityId,
+	trx = db,
+}: {
+	pubId: PubsId;
+	relations: Record<string, Omit<AddPubRelationsInput, "slug">[]>;
+	communityId: CommunitiesId;
+	trx?: typeof db;
+}) => {
+	if (!Object.keys(relations).length) {
+		return;
+	}
+
+	await maybeWithTrx(trx, async (trx) => {
+		const slugs = Object.keys(relations);
+
+		await removeAllPubRelationsBySlugs({ pubId, slugs, communityId, trx });
+
+		const relationsWithSlug = slugs.flatMap((slug) =>
+			relations[slug].map((relation) => ({ ...relation, slug }) as AddPubRelationsInput)
+		);
+
+		await upsertPubRelations({ pubId, relations: relationsWithSlug, communityId, trx });
+	});
+};
+
+export const updatePub = async ({
```

Comment:
i did change how this works, see individual commits/compare with old _updatePub. mainly the upserts are not by id, rely on the unique index